[MTV-1794] Update provider vCenter validation

jpuzz0 opened the pull request:
## 📝 Links
https://issues.redhat.com/browse/MTV-1794

## 📝 Description
Address feedback related to this comment: https://issues.redhat.com/browse/MTV-1794?focusedId=26727397&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-26727397

## 🎥 Demo
https://github.com/user-attachments/assets/5d7e8a5e-52db-48b0-9e21-b47de63b0600

avivtur commented:
/lgtm

sgratch commented on packages/forklift-console-plugin/src/modules/Providers/utils/validators/provider/vsphere/validateVCenterURL.ts:
```diff
@@ -53,18 +57,21 @@ export const validateVCenterURL = (url: string, caCert?: string): ValidationMsg
       return {
         type: 'error',
-        msg: 'Invalid URL. The URL must be a fully qualified domain name (FQDN).',
+        msg: 'Invalid URL. The URL must be a fully qualified domain name (FQDN) and match the FQDN in the certificate you uploaded.',
```
I would avoid validating `urlMatchesCertFqdn` and blocking with an error due to the following reasons:
1. It is validated for fqdn urls and we don't really need that since in case of an invalid not matched url/cacert, the connection will fail with an alert error. Adding this validation for both the url might be problematic since blocking the user and I'm not sure it's always right. We can always display a warning but I think we don't need that for fqdn. 
2. Even if we do decide to add a warning for such a case,  this error should be displayed on the cacert or on the url fields? not sure. If url is valid and an invalid not matched cacert was set then why not displaying that on the cacert field?
3. When switching to ESXi end point, no such validation is done on a fqdn url, which is confusing.
4. When using ip address for the url, a warning is displayed and not an error in case of a not matched cacert - confusing
5. When editing a cacert or a url fields for an existing provider, no such validation is done.

Therefore I suggest that instead of making the code complicated  with that validation logic based on 2 fields, just display a warning only if the url includes an ip - as done in lines 71-75 (we can verify the ip within the cacert as long as it's a warning and not not an error  but not sure it's needed).  
This is the requirement of the bug and adding more validations might be risky since blocking and we have the backend for that.

sgratch reviewed (commented):
@jpuzz0 please see comment below
I suggest removing the `urlMatchesCertFqdn` ui validation and leave only the following validation:

```
f (isValidIpAddress) {
      return {
        type: 'warning',
        msg: 'The URL is not a fully qualified domain name (FQDN). If the certificate does not match the URL, the connection might fail.',
};
```

sgratch commented on packages/forklift-console-plugin/src/modules/Providers/utils/validators/provider/vsphere/validateVCenterURL.ts:
```diff
@@ -53,18 +57,21 @@ export const validateVCenterURL = (url: string, caCert?: string): ValidationMsg
-        type: 'error',
-        msg: 'Invalid URL. The URL must be a fully qualified domain name (FQDN) and match the FQDN in the certificate you uploaded.',
+        type: 'warning',
+        msg: 'The URL is not a fully qualified domain name (FQDN). If the certificate does not match the URL, the connection might fail.',
```
I would avoid validating `urlMatchesCertFqdn` and blocking with an error due to the following reasons:
1. It is validated for fqdn urls and we don't really need that since in case of an invalid not matched url/cacert, the connection will fail with an alert error. Adding this validation for both the url might be problematic since blocking the user and I'm not sure it's always right. We can always display a warning but I think we don't need that for fqdn. 
2. Even if we do decide to add a warning for such a case,  this error should be displayed on the cacert or on the url fields? not sure. If url is valid and an invalid not matched cacert was set then why not displaying that on the cacert field?
3. When switching to ESXi end point, no such validation is done on a fqdn url, which is confusing.
4. When using ip address for the url, a warning is displayed and not an error in case of a not matched cacert - confusing
5. When editing a cacert or a url fields for an existing provider, no such validation is done.

Therefore I suggest that instead of making the code complicated  with that validation logic based on 2 fields, just display a warning only if the url includes an ip - as done in lines 71-75 (we can verify the ip within the cacert as long as it's a warning and not not an error  but not sure it's needed).  
This is the requirement of the bug and adding more validations might be risky since blocking and we have the backend for that.

jpuzz0 replied:
Sure, I'll just keep the warning and remove that FQDN check. 

1. I was under the impression the backend was not sophisticated enough to catch FQDN mismatches - speaking to your first comment. 

2. The error was being displayed under both fields (URL and CA cert) when triggered, which I think was appropriate. 

4. I don't think it's confusing the check for a FQDN only when the URL is a name and not an IP address, but that's just my opinion.

3, 5. I agree the inclusion of this validation other places would make more sense, so since that's very much out of scope, I'll just remove as requested.

jpuzz0 replied:
Updated

sgratch commented on packages/forklift-console-plugin/src/modules/Providers/utils/validators/provider/providerValidator.ts:
```diff
@@ -27,7 +27,7 @@ export function providerValidator(
       break;
     case 'vsphere':
-      validationError = vsphereProviderValidator(provider, subType, secret?.data?.cacert);
+      validationError = vsphereProviderValidator(provider, subType, secret);
```
do we still need to pass the whole secret instead of just the cacert in all relevant places?

jpuzz0 replied:
The cacert isn't read anymore for the validation, just the const insecureSkipVerify field. I could just pass that if you'd like, but I'm not sure it makes much of a difference here.

jpuzz0 replied:
Updated

sgratch commented:
CC:// @RichardHoch @anarnold97

jpuzz0 commented:
> @jpuzz0 Thanks for removing the error. Please check the following comments:
> 
> 1. Still need to handle the edit URL for vcenter verification since currently when editing a URL of an existing vcenter insecure provider, the warning of "The URL is not a fully qualified domain name (FQDN)..."  is initially displayed:
> 
>  Screencast.from.2025-03-18.16-12-35.webm 
> 2. Consider rephrasing the warning text to something like:
>    "The URL is not a fully qualified domain name (FQDN). If the certificate is not skipped and does not match the URL, the connection might fail. "
> 
> @RichardHoch can you please advice?

@sgratch I've updated the text to your suggestion and updated the edit validation.

sgratch reviewed (commented):
All issues mentioned above were addresses. 
I see that something seems broken with the `VSphereEditURLModal` url field validation for the vcenter end point. The validation text message is not re-rendered as required and the initial value is also wrong (note that the text message  flickers once opening the modal):

Before this fix:

[Screencast from 2025-03-19 12-07-59.webm](https://github.com/user-attachments/assets/535c97c1-b3b5-4f5b-bb9c-0be0303ead48)

After this fix:

[Screencast from 2025-03-19 12-09-20.webm](https://github.com/user-attachments/assets/95be7d2e-4ab8-4b93-816d-dfcd24da484f)

jpuzz0 commented:
@sgratch The flickering is due to us now having to fetch the secret to make this validation check  - since it depends on whether the connection is marked as secure or not. I will make the request happen on page entry versus modal open to prevent this. 

Otherwise, I'm not sure what the issue is. The error message you're experiencing in the first video doesn't exist anymore because you told me to remove it. And the warning will only show when the `Skip certificate validation` switch is off.
